### DISCUSSION TITLE
Add cron support to Schedules

### DISCUSSION THREAD

**Discussion by mal-drop** [nitrictech/nitric#327] [Ideas]
We have a requirement to be able schedule notification emails to users at 7 am every Tuesday.  We are currently implementing this feature on a Node service using node-cron (https://www.npmjs.com/package/node-cron), and specifying the local timezone 'Australia/Sydney'.

Please add cron expression support for Nitric Scheduler on AWS, ideally also add locale support as well.

```javascript
import { schedule } from '@nitric/sdk';

// Create a schedule that runs Tuesday at 7 AM
schedule('send-reminder').locale('Australia/Sydney').cron('0 7 * * 1', async (ctx) => {
  // Code which sends a reminder
});
```

We currently have 8 cron schedule in our system, 7 of which are genuine cron jobs and 1 which could use a simple timer. This number will grow as we deliver additional functionality.

**Comment by raksiv** [comment]
Had a chat with @jyecusch about this one, looks like we support it at the membrane level, but not at the SDK levels.

There are some implementation differences between the clouds
AWS - cron(Minutes Hours Day-of-month Month Day-of-week Year)
GCP - cron(Minutes Hours Day-of-month Month Day-of-week)
Azure - {second} {minute} {hour} {day} {month} {day-of-week}

**Reply by mal-drop** [reply]
Good to hear its available in Membrane. I think documenting the differences between the clouds would be adequate, its much better to have this feature than not because of some differences between the clouds.

Looks like there is also cron trigger support in Azure:
https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-timer?tabs=in-process&pivots=programming-language-csharp

**Comment by raksiv** [comment]
Thanks, agree with your comments about documenting vs omitting.

I've updated the above comment to include Azure, I'll work with the team to convert this into a story and get it prioritized.

**Comment by tjholm** [comment]
After some research on this, we can expose the cron expressions very easily just need to add the interface method to the SDKs.

Adding TZ support is a little bit trickier, as in the cases of AWS we'll need to write a conversion tool to shift the CRON expression relative to UTC (as UTC in the only timezone AWS supports). I've found some libraries in non-golang languages which can do this so can look at porting those to golang and open-source this :).

I think we can do this in two different steps, just to get this out a bit quicker:

Feature 1. Expose just the cron expression setting, with the default TZ always being UTC, hours in the cronjob can be adjusted relative to UTC. e.g. 10AM every day in UTC -> `0 10 * * *` could be translated to 11PM every day in UTC (this would be 10AM AEDT) `0 23 * * *`

Feature 2. Add Timezone setting for both the rate and cron expressions to allow pinning them to a specific timezone and we can adjust the final CRON expression accordingly. The main issue with this is daylight saving and documenting behaviour around that. This will also be backwards compatible with Feature 1 as UTC will always be the default timezone.

**Reply by mal-drop** [reply]
Go to love the person who invented day light savings. I know the first person that will be visited when time travel is invented.

**Reply by tjholm** [reply]
But will you arrive on time to get them or an hour too early or late? who knows! :confused:

**Comment by tjholm** [comment]
Have raised https://github.com/nitrictech/node-sdk/pull/144 to add initial cron support for the node SDK (this is already supported in the CLI so should be ready to test once it's merged into develop).
